### PR TITLE
Add feature to view past elections

### LINKED ISSUE #45: Display results on page
Is it possible to display the results of the previous elections on the page too?

### PR DESCRIPTION
**Pull request by LMBishop**
Closes #45 

Two new API endpoints

- `GET /api/election/results`
- `POST /api/admin/election/publish`

And an updated database schema

image

Election results are no longer presented on `/stats`, rather the user is directed straight to the results page (`/results/[eid]`) when an election is ended

### CODE REVIEW

**Comment by MattyTheHacker** [comment]
Is there a method of unpublishing results?

**Comment by LMBishop** [comment]
Yes

**Comment by MattyTheHacker** [comment]
epic

**Comment by Thatsmusic99** [comment]
Just another quick question regarding results as well - is the same level of detail available to everyone when published, e.g. vote counts for a certain candidate?

**Comment by LMBishop** [comment]
Yes

File: internal/httpcore/endpoints_api_self.go
Comment (codemicro) [inline internal/httpcore/endpoints_api_self.go:74]: 🤦🏼‍♀️

File: internal/instantRunoff/instantRunoff.go
Comment (codemicro) [inline internal/instantRunoff/instantRunoff.go:81]: This isn't logging any frequencies now, it's just tallying votes so it's probably worth changing this function's name. Since it's also not messing with any variables in `Run`'s scope any more, it's worth pulling this whole function out into it's own top level thing.

File: internal/instantRunoff/instantRunoff.go
Comment (codemicro) [inline internal/instantRunoff/instantRunoff.go:137]: I'd pull this one level up so it's not in the same scope as n, since the only reason that this block is its own block is to show that the whole thing from line 126 to 138 is one operation, ie. filter ids.

File: internal/httpcore/endpoints_elections.go
Comment (codemicro) [inline internal/httpcore/endpoints_elections.go:447]: Instead of `election_id`, I would be tempted to use `id` to keep consistency with the rest of the codebase where only one type of ID is involved, eg. in `apiAdminStartElection`

File: web/src/routes/results/[eid]/+page.ts
Comment (codemicro) [inline web/src/routes/results/[eid]/+page.ts:13]: This needs to be a GET request

**Review by codemicro — CHANGES_REQUESTED** [review]
This review is only for the backend stuff - I've left the detailed review of the frontend things to @Vishwas-Adiga if he wants to review them, but I've highlighted a bug that I found that stopped me testing it all.

Overall, looks good! I'm really glad that the instant runoff module now returns a concrete data type instead of just a big ol' string and I'm glad it's got some unit tests. Those're both things that I wanted to do in the first instance but didn't because of time pressure and promptly then never did.

I'm a bit iffy about the election results being entirely public. As committee, you guys get the final say on this, but the system was designed to stop the runners-up ranking being shown to anyone who's not committee, and the vote count being shown to anybody. The latter being shown to committee is no big deal, but the former is definitely a departure from what I was trying to keep to when I was originally writing socvot, so it might be worth considering if you want that to be public knowledge or not.

**Comment by codemicro** [comment]
Once that bug has been fixed, I'll do a proper test

**Comment by LMBishop** [comment]
> I'm a bit iffy about the election results being entirely public. As committee, you guys get the final say on this, but the system was designed to stop the runners-up ranking being shown to anyone who's not committee, and the vote count being shown to anybody. The latter being shown to committee is no big deal, but the former is definitely a departure from what I was trying to keep to when I was originally writing socvot, so it might be worth considering if you want that to be public knowledge or not.

I suppose this is a discussion to be had within the committee, although we already publish full election results in the EGM/AGM minutes (with tallies and all), so I don't see much of a difference here. I also personally think it is better for the results to be public for the sake of transparency.

**Comment by codemicro** [comment]
Do we definitely publish those minutes? I can't find a copy of them anywhere online - but yes, committee internal discussion required. It's not my decision to make

**Comment by codemicro** [comment]
Proper test all looks good to me, happy to approve once the rest of the comments have been dealt with/rejected

**Comment by LMBishop** [comment]
> Do we definitely publish those minutes?

If they weren't recently then they should've been!

File: internal/instantRunoff/instantRunoff.go
Comment (LMBishop) [reply internal/instantRunoff/instantRunoff.go]: This is now `getTallies` and is a top level function

File: web/src/routes/results/[eid]/+page.svelte
Comment (Vishwas-Adiga) [inline web/src/routes/results/[eid]/+page.svelte:37]: Would be nicer to use DateTimeFormat, something like so:
```
$: date = new Intl.DateTimeFormat("en-GB", {
		dateStyle: "full",
		timeStyle: "long",
	}).format(new Date(electionOutcome.date));
```

This would format the date as `Sunday 28 July 2024 at 20:32:30 BST`, which reads better with the "at" separating the date and time IMO.

File: web/src/routes/results/[eid]/+page.svelte
Comment (Vishwas-Adiga) [inline web/src/routes/results/[eid]/+page.svelte:87]: Not sure if this wrapper div is needed? See comment below about using `grid` instead

File: web/src/routes/results/[eid]/+page.svelte
Comment (Vishwas-Adiga) [inline web/src/routes/results/[eid]/+page.svelte:101]: Can be reworded to "This election was held on <b>..." if using DateTimeFormat as suggested above

File: web/src/routes/results/[eid]/+page.svelte
Comment (Vishwas-Adiga) [inline web/src/routes/results/[eid]/+page.svelte:147]: Seeing as this is the most important bit of info on the page, IMO it should be at the top of the panel

File: web/src/routes/results/[eid]/+page.svelte
Comment (Vishwas-Adiga) [inline web/src/routes/results/[eid]/+page.svelte:222]: Probably worth switching to a grid layout seeing as it's consistently used across the project. Should be a simple change that also eliminates the need for the `.button` class.

```
	div.unpublished-callout {
		display: grid;
		grid-template-columns: 1fr auto;
		align-items: center;
		gap: 16px;
	}
```

File: web/src/routes/results/[eid]/+page.svelte
Comment (Vishwas-Adiga) [inline web/src/routes/results/[eid]/+page.svelte:241]: Not sure what this targets? We might want to use a more specific selector or get rid of it altogether

File: web/src/routes/results/[eid]/+page.svelte
Comment (Vishwas-Adiga) [inline web/src/routes/results/[eid]/+page.svelte:243]: Table text is rendered in Times New Roman (on my device at least (Chrome on Windows)) which I suspect comes from the UA stylesheet. Definitely worth overriding this with sans-serif or Roboto Mono either here or in static/globals.css.

Comment (Vishwas-Adiga) [reply web/src/routes/results/[eid]/+page.svelte:243]: image

File: web/src/routes/results/[eid]/+page.svelte
Comment (Vishwas-Adiga) [inline web/src/routes/results/[eid]/+page.svelte:111]: Rename to `Outcome`?

**Review by Vishwas-Adiga — CHANGES_REQUESTED** [review]
Haven't looked at the go stuff but the front-end looks mostly good to go, barring a few minor bits.

File: web/src/routes/results/[eid]/+page.svelte
Comment (LMBishop) [reply web/src/routes/results/[eid]/+page.svelte:147]: Maybe, though in my mind the page is supposed to be read as a sequence of events in order (like it is being announced)

File: web/src/routes/results/[eid]/+page.svelte
Comment (LMBishop) [reply web/src/routes/results/[eid]/+page.svelte:243]: Oh gross, for reference it is supposed to use the same font as the rest of the page
screenshot
(FF on Linux)
I guess I should really install some actual fonts on my system...

**Review by codemicro — APPROVED** [review]
Backend looks a-ok :)

File: web/src/routes/results/[eid]/+page.svelte
Comment (Vishwas-Adiga) [reply web/src/routes/results/[eid]/+page.svelte:148]: Ok, that's fair 👍

**Comment by codemicro** [comment]
@Vishwas-Adiga do you think this is ready to merge now?

**Comment by Vishwas-Adiga** [comment]
Yes, everything LGTM
I was just waiting on @LMBishop to re-request a review once he's happy with his changes

**Review by Vishwas-Adiga — APPROVED** [review]

**Comment by codemicro** [comment]
We merging this?

**Comment by codemicro** [comment]
sounds like it's time to merge

**Pull request merged**
